### DISCUSSION TITLE
Using AWS lambda response streaming with Sveltekit

### DISCUSSION THREAD

**Discussion by H0R5E** [sveltejs/kit#9809] [Q&A]
Hi, I am developing a [SvelteKit adapter for AWS](https://github.com/Data-Only-Greater/sveltekit-adapter-aws-base) (which can be used with different IAC providers).

When I started building the adapter, AWS Lambda was not capable of generating a streamed response, so using [SvelteKit's streaming features](https://svelte.dev/blog/streaming-snapshots-sveltekit) wasn't possible.

At the beginning of April, AWS announced [AWS Lambda response streaming](https://aws.amazon.com/blogs/compute/introducing-aws-lambda-response-streaming/), with a basic example of a handler like so:

```node
exports.handler = awslambda.streamifyResponse(
    async (event, responseStream, context) => {
        responseStream.setContentType(“text/plain”);
        responseStream.write(“Hello, world!”);
        responseStream.end();
    }
);
```

Alternatively, they suggest using a `pipeline` as so:

```node
const pipeline = require("util").promisify(require("stream").pipeline);
const zlib = require('zlib');
const { Readable } = require('stream');

exports.gzip = awslambda.streamifyResponse(async (event, responseStream, _context) => {
    // As an example, convert event to a readable stream.
    const requestStream = Readable.from(Buffer.from(JSON.stringify(event)));
    
    await pipeline(requestStream, zlib.createGzip(), responseStream);
});
```

My adapter currently handles the request/response handling something like so:

```node
import { Server } from '../index.js'
import { manifest } from '../manifest.js'

export async function handler(event, context) {
  const app = new Server(manifest)
  const {
    rawPath,
    headers,
    rawQueryString,
    body,
    requestContext,
    isBase64Encoded,
    cookies,
  } = event

  // Some manipulation here

  const request = new Request(rawURL, {
    method: requestContext.http.method,
    headers: new Headers(headers),
    body: rawBody,
  })

  const rendered = await app.respond(request, {
    platform: { context },
  })

  // Parse the response into lambda proxy response
  let response

  if (rendered) {
    response = {
      // some header handling here
      body: await rendered.text(),
      statusCode: rendered.status,
    }
  } else {
    response = {
      statusCode: 404,
      body: 'Not found.',
    }
  }

  return response
}
```

I was just wondering if anyone had any advice on modifying the adapter to use the AWS response stream features, before I need to take a deep dive?

Many thanks for your help,

Mat

**Comment by H0R5E** [comment]
Having had time for a look around, I think I can leverage the [`setResponse`](https://github.com/sveltejs/kit/blob/39f1631417014aff279d6073b4c1480682484685/packages/kit/src/exports/node/index.js#L107) function here. So it might be as easy as:

```node
import { setResponse } from '@sveltejs/kit/node';

exports.handler = awslambda.streamifyResponse(
    async (event, responseStream, context) => {
        const app = new Server(manifest)
        const {
          rawPath,
          headers,
          rawQueryString,
          body,
          requestContext,
          isBase64Encoded,
          cookies,
        } = event

        // Some manipulation here

         const request = new Request(rawURL, {
         method: requestContext.http.method,
          headers: new Headers(headers),
          body: rawBody,
        })

        const rendered = await app.respond(request, {
          platform: { context },
        })
        setResponse(responseStream, rendered)
    }
)
```

**Comment by GeoffreyBooth** [comment]
It would seem that @geoffrich’s article https://svelte.dev/blog/streaming-snapshots-sveltekit needs an update.

> SvelteKit will only be able to stream responses if your app’s hosting platform supports it. In general, any platform built around AWS Lambda (e.g. serverless functions) will not support streaming, but any traditional Node.js server or edge-based runtime will.

This would appear to be no longer accurate, per https://docs.aws.amazon.com/lambda/latest/dg/configuration-response-streaming.html.

I also think that AWS lambda is a significant enough platform to justify a first-party SvelteKit adapter.

**Reply by ivanbarlog** [reply]
I took a fresh approach here https://www.npmjs.com/package/kit-on-lambda

It is still in beta and I am trying to collect some feedback on it so if you are interested you can try it and let me know what you think.
